# Loading Data

In [32]:
# Import libraries
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

In [33]:
# Load the dataset
dataset = pd.read_csv("../../Data/Churn Prediction Data/churn-bigml-20.csv")

print(dataset.head())

  State  Account length  Area code International plan Voice mail plan  \
0    LA             117        408                 No              No   
1    IN              65        415                 No              No   
2    NY             161        415                 No              No   
3    SC             111        415                 No              No   
4    HI              49        510                 No              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                      0              184.5               97   
1                      0              129.1              137   
2                      0              332.9               67   
3                      0              110.4              103   
4                      0              119.3              117   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             31.37              351.6               80             29.89   
1             21.95   

# Looking for dataset information

In [34]:
# All details of the dataset
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 667 entries, 0 to 666
Data columns (total 20 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   State                   667 non-null    object 
 1   Account length          667 non-null    int64  
 2   Area code               667 non-null    int64  
 3   International plan      667 non-null    object 
 4   Voice mail plan         667 non-null    object 
 5   Number vmail messages   667 non-null    int64  
 6   Total day minutes       667 non-null    float64
 7   Total day calls         667 non-null    int64  
 8   Total day charge        667 non-null    float64
 9   Total eve minutes       667 non-null    float64
 10  Total eve calls         667 non-null    int64  
 11  Total eve charge        667 non-null    float64
 12  Total night minutes     667 non-null    float64
 13  Total night calls       667 non-null    int64  
 14  Total night charge      667 non-null    fl

In [35]:
# Various measures of the numerical data
print(dataset.describe())

       Account length   Area code  Number vmail messages  Total day minutes  \
count      667.000000  667.000000             667.000000         667.000000   
mean       102.841079  436.157421               8.407796         180.948126   
std         40.819480   41.783305              13.994480          55.508628   
min          1.000000  408.000000               0.000000          25.900000   
25%         76.000000  408.000000               0.000000         146.250000   
50%        102.000000  415.000000               0.000000         178.300000   
75%        128.000000  415.000000              20.000000         220.700000   
max        232.000000  510.000000              51.000000         334.300000   

       Total day calls  Total day charge  Total eve minutes  Total eve calls  \
count       667.000000        667.000000         667.000000       667.000000   
mean        100.937031         30.761769         203.355322       100.476762   
std          20.396790          9.436463        

Interpretation of the information of the above info data
- `<class 'pandas.core.frame.DataFrame'>` - shows the type of the dataset
- Rows - 667
- Cols - 20
- Shape - `(667,20)`
- How many cols of each type?
    - `bool(1), float64(8), int64(8), object(3)`
    - This helps in identification of **Categorical** cols.
    - Categorical cols - `State`, `International plan`,`Voice mail plan` and `Churn`.

# Handling missing and duplicated values

In [36]:
# Checking for missing values
print(dataset.isna().sum())

State                     0
Account length            0
Area code                 0
International plan        0
Voice mail plan           0
Number vmail messages     0
Total day minutes         0
Total day calls           0
Total day charge          0
Total eve minutes         0
Total eve calls           0
Total eve charge          0
Total night minutes       0
Total night calls         0
Total night charge        0
Total intl minutes        0
Total intl calls          0
Total intl charge         0
Customer service calls    0
Churn                     0
dtype: int64


In [37]:
# Checking for duplicated data
print(dataset.duplicated().sum())

0


Zero missing and duplicated values. Good to go!

# Categorical Cols

Categorical cols - `State`, `International plan`,`Voice mail plan` and `Churn`.

Columns like `International plan` and `Voice mail plan` has `Yes` and `No` as values which is object type which gets converted to `1` for `Yes` and `0` for `No`.

In [38]:
# Converting categorical data to numerical data
dataset['International plan'] = dataset['International plan'].map({'Yes': 1, 'No': 0})
dataset['Voice mail plan'] = dataset['Voice mail plan'].map({'Yes': 1, 'No': 0})

print(dataset[['International plan','Voice mail plan']].head())

   International plan  Voice mail plan
0                   0                0
1                   0                0
2                   0                0
3                   0                0
4                   0                0


`Churn` column is boolean which can be replaced by 0 and 1.

In [39]:
# Converting target column from boolean to int data type
dataset['Churn'] = dataset['Churn'].astype(int)
print(dataset[['Churn']].head())

   Churn
0      0
1      1
2      1
3      0
4      0


`State` column is unable to represent in order and has multiple entries which leads to `one-hot encoding`.

In [40]:
# One hot encoding for 'State' column and converting that to into int data type
dataset = pd.get_dummies(dataset, columns=['State'], drop_first=True).astype(int)

# Features Selection

Rearranging columns to seperate features easily

In [41]:
# Rearranging columns
cols = ["Churn"]
dataset = dataset[cols + [col for col in dataset.columns if col not in cols]]
print(dataset.head())

   Churn  Account length  Area code  International plan  Voice mail plan  \
0      0             117        408                   0                0   
1      1              65        415                   0                0   
2      1             161        415                   0                0   
3      0             111        415                   0                0   
4      0              49        510                   0                0   

   Number vmail messages  Total day minutes  Total day calls  \
0                      0                184               97   
1                      0                129              137   
2                      0                332               67   
3                      0                110              103   
4                      0                119              117   

   Total day charge  Total eve minutes  ...  State_SD  State_TN  State_TX  \
0                31                351  ...         0         0         0   
1   

In [42]:
# Conerting data ready for features selection
data = dataset.values

In [43]:
# Features selection
X = data[:,1:]
y = data[:, 0]

Spliting of Dataset - Train/Test set

In [44]:
# Spliting of data into training and testing test
# 80/20 ratio
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

Feature scaling

In [45]:
# Feature scaling
scaler = StandardScaler()

# Fitting and scaling X_train
X_train = scaler.fit_transform(X_train)

# Scaling X_test
X_test = scaler.transform(X_test)

In [46]:
# Printing shapes for splited data to confirm the spliting

print("X_train shape = ", X_train.shape)
print("X_test shape = ", X_test.shape)
print("y_train shape = ", y_train.shape)
print("y_test shape = ", y_test.shape)

X_train shape =  (533, 68)
X_test shape =  (134, 68)
y_train shape =  (533,)
y_test shape =  (134,)
